collect_app/src/main/java/org/odk/collect/android/activities/FormDownloadList.java

```diff
@@ -665,42 +674,151 @@ public void formListDownloadingComplete(HashMap<String, FormDetails> result) {
+        ArrayList<String> ids = new ArrayList<String>(forms.keySet());
+        for (int i = 0; i < forms.size(); i++) {
+            String formDetailsKey = ids.get(i);
+            FormDetails details = forms.get(formDetailsKey);
+            HashMap<String, String> item = new HashMap<String, String>();
+            item.put(FORMNAME, details.formName);
+            item.put(FORMID_DISPLAY,
+                    ((details.formVersion == null) ? "" : (getString(R.string.version) + " " + details.formVersion + " ")) +
+                            "ID: " + details.formID );
+            item.put(FORMDETAIL_KEY, formDetailsKey);
+            item.put(FORM_ID_KEY, details.formID);
+            item.put(FORM_VERSION_KEY, details.formVersion);
+
+            // Insert the new form in alphabetical order.
+            if (mFormList.size() == 0) {
+                mFormList.add(item);
+            } else {
+                int j;
+                for (j = 0; j < mFormList.size(); j++) {
+                    HashMap<String, String> compareMe = mFormList.get(j);
+                    String name = compareMe.get(FORMNAME);
+                    if (name.compareTo(forms.get(ids.get(i)).formName) > 0) {
+                        break;
+                    }
+                }
+                mFormList.add(j, item);
+            }
+        }
+        selectSupersededForms();
+        mFormListAdapter.notifyDataSetChanged();
+    }
+
+    public static HashMap<String, HashMap<String, FormDetails>> splitFormsIntoAccounts(
+            Context context, final HashMap<String, FormDetails> forms) {
+
+        HashMap<String, HashMap<String, FormDetails>> result = new HashMap<>();
+        result.put(context.getResources().getString(R.string.all_forms), forms);
+
+        SharedPreferences settings =
+                PreferenceManager.getDefaultSharedPreferences(context);
+        boolean showSharedForms = settings.getBoolean(PreferencesActivity.KEY_SHOW_SHARED_FORMS,
+                true);
+        String username = settings.getString(PreferencesActivity.KEY_USERNAME, null);
+
+        for(FormDetails curForm : forms.values()) {
+            String user = getOnaUser(curForm.downloadUrl);
+            if(user != null) {
+                //check if user wants to see shared forms
+                if(username != null && !showSharedForms && !user.equals(username)) {
+                    continue;
+                }
+
+                String userKey = context.getResources().getString(R.string
+                        .forms_by, user);
+                if(!result.containsKey(userKey)) {
+                    result.put(userKey,
+                            new HashMap<String, FormDetails>());
+                }
+
+                result.get(userKey).put(curForm.formID, curForm);
+            } else {
+                Log.e(t, "Could not extract the user that owns the specified form");
+            }
+        }
+
+        if(result.size() == 2) {//forms from only one account
```

Comment:
All forms are initially all put in the first item in the map (called "all_forms"). Cases where you end up with two items in the map after splitting forms by accounts means that there's only one account